packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -0,0 +1,1377 @@
+        val file = uriToFile(fileUri)
+        promise.resolve(contentUriFromFile(file).toString())
+      } else {
+        promise.reject("ERR_FILESYSTEM_CANNOT_READ_DIRECTORY", "No readable files with the uri: $uri. Please use other uri.")
+      }
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject(e)
+    }
+  }
+
+  private fun contentUriFromFile(file: File): Uri {
+    return try {
+      val activityProvider: ActivityProvider by moduleRegistry()
+      val application = activityProvider.currentActivity.application
+      FileProvider.getUriForFile(application, application.packageName + ".FileSystemFileProvider", file)
+    } catch (e: Exception) {
+      throw e
+    }
+  }
+
+  @ExpoMethod
+  fun readSAFDirectoryAsync(uriStr: String?, options: Map<String?, Any?>?, promise: Promise) {
+    try {
+      val uri = Uri.parse(uriStr)
+      ensurePermission(uri, Permission.READ)
+      if (isSAFUri(uri)) {
+        val file = DocumentFile.fromTreeUri(context, uri)
+        if (file == null || !file.exists() || !file.isDirectory) {
+          promise.reject(
+            "ERR_FILESYSTEM_CANNOT_READ_DIRECTORY",
+            "Uri '$uri' doesn't exist or isn't a directory."
+          )
+          return
+        }
+        val children = file.listFiles()
+        val result: MutableList<String> = ArrayList()
+        for (child in children) {
+          result.add(child.uri.toString())
+        }
+        promise.resolve(result)
+      } else {
+        throw IOException("The URI '$uri' is not a Storage Access Framework URI. Try using FileSystem.readDirectoryAsync instead.")
+      }
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject(e)
+    }
+  }
+
+  @ExpoMethod
+  fun makeSAFDirectoryAsync(uriStr: String?, dirName: String?, promise: Promise) {
+    try {
+      val uri = Uri.parse(uriStr)
+      ensurePermission(uri, Permission.WRITE)
+      if (isSAFUri(uri)) {
+        val dir = getNearestSAFFile(uri)
+        if (dir != null) {
+          if (!dir.isDirectory) {
+            promise.reject("ERR_FILESYSTEM_CANNOT_CREATE_DIRECTORY", "Provided uri '$uri' is not pointing to a directory.")
+            return
+          }
+        }
+        val newDir = dirName?.let { dir?.createDirectory(it) }
+        if (newDir == null) {
+          promise.reject("ERR_FILESYSTEM_CANNOT_CREATE_DIRECTORY", "Unknown error.")
+          return
+        }
+        promise.resolve(newDir.uri.toString())
+      } else {
+        throw IOException("The URI '$uri' is not a Storage Access Framework URI. Try using FileSystem.makeDirectoryAsync instead.")
+      }
+    } catch (e: Exception) {
+      promise.reject(e)
+    }
+  }
+
+  @ExpoMethod
+  fun createSAFFileAsync(uriStr: String?, fileName: String?, mimeType: String?, promise: Promise) {
+    try {
+      val uri = Uri.parse(uriStr)
+      ensurePermission(uri, Permission.WRITE)
+      if (isSAFUri(uri)) {
```

Comment:
Invert the if flow. This will decrease nesting a bit
```kotlin
if (!isSAFUri(uri)) {
  throw IOException();
}
// here rest of the flow under that if
```